modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+
+/*
+  Below are segment statuses, which is a way to make visible to outsiders the
+  current ongoing operation.
+*/
+private const STATUS_UNLOCKED : uint = 0;
+private const STATUS_ADD : uint = 1;
+private const STATUS_REMOVE : uint = 2;
+private const STATUS_LOOKUP : uint = 3;
+private const STATUS_BALANCE : uint = 4;
+
+/*
+  Below are statuses specific to the work stealing algorithm.
+*/
+private const WS_INITIALIZED : int = -1;
+private const WS_FINISHED_WITH_NO_WORK : int = 0;
+private const WS_FINISHED_WITH_WORK : int = 1;
+
+/*
+  Frozen states... If we are FREEZE_UNFROZEN, we are mutable. If we are FREEZE_FROZEN,
+  we are immutable. If we are FREEZE_MARKED, we are in the middle of a state change.
+*/
+private const FREEZE_UNFROZEN = 0;
+private const FREEZE_MARKED = 1;
+private const FREEZE_FROZEN = 2;
+
+/*
+  The phases for operations.
+*/
+private const ADD_BEST_CASE = 0;
+private const ADD_AVERAGE_CASE = 1;
+private const REMOVE_BEST_CASE = 2;
+private const REMOVE_AVERAGE_CASE = 3;
+private const REMOVE_WORST_CASE = 4;
+
+/*
+  The initial amount of elements in an unroll block. Each successive unroll block
+  is double the size of it's predecessor.
+*/
+config param INITIAL_BLOCK_SIZE = 1024;
+/*
+  To prevent stealing too many elements (horizontally) from another node's segment
+  (hence creating an artifical load imbalance), if the other node's segment has
+  less than a certain threshold (see `WORK_STEALING_MAX_MEMORY_MB`) but above
+  another threshold (see `WORK_STEALING_MINIMUM`), we steal a percentage of their
+  elements, leaving them with majority of their elements. This way, the amount the
+  other segment loses is proportional to how much it owns, ensuring a balance.
+*/
+config param WORK_STEALING_RATIO = 0.25;
+/*
+  The maximum amount of work to steal from a horizontal node's segment. This
+  should be set to a value, in megabytes, that determines the maximum amount of
+  data that should be sent in bulk at once. The maximum number of elements is
+  determined by: (WORK_STEALING_MAX_MEMORY_MB * 1024 * 1024) / sizeof(eltType).
+  For example, if we are storing 8-byte integers and have a 1MB limit, we would
+  have a maximum of 125,000 elements stolen at once.
+*/
+config param WORK_STEALING_MAX_MEMORY_MB : real = 1.0;
+/*
+  Determines whether or not we check if we are 'frozen'. These checks cause a
+  significant decline in performance, of about 33%. If the utility is not needed,
+  as it is more useful when reads are performed more often than writes, it may
+  be disabled for a raw performance boost. If the user attempts to freeze the
+  bag with this set to true, it merely returns false; all other checks are elided.
+*/
+config param BAG_NO_FREEZE = false;
+/*
+  The minimum number of elements a horizontal segment must have to become eligible
+  to be stolen from. This may be useful if some segments produce less elements than
+  others and should not be stolen from.
+*/
+config param WORK_STEALING_MINIMUM = 1;

```

Comment:
I can agree to the prefixing and camel case. I'll hold off on this in case anyone else want to give their input.